provider/openstack/errors.go

```diff
@@ -0,0 +1,29 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package openstack
+
+import (
+	gooseerrors "gopkg.in/goose.v2/errors"
+
+	"github.com/juju/juju/environs/context"
+)
+
+func MaybeHandleCredentialError(err error, ctx context.ProviderCallContext) (error, bool) {
+	isUnauthorized := gooseerrors.IsUnauthorised(err)
+	//fmt.Printf("XXXXX... IsUnauthorized : %v\n", IsUnauthorized)
+	if ctx != nil && isUnauthorized {
+		invalidateErr := ctx.InvalidateCredential("openstack cloud denied access")
+		if invalidateErr != nil {
+			logger.Warningf("could not invalidate stored openstack cloud credential on the controller: %v", invalidateErr)
+		}
+	}
+	return err, isUnauthorized
```

Comment:
Do we actually need to use the isAuthorised bool anywhere?
We do in maas but why do we need it here?
Would a better approach be to use what's done with ec2 provider with the maybeConvertCredentialError() method?